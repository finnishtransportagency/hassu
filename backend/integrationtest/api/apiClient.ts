import { AbstractApi, OperationConfig } from "../../../common/abstractApi";
import { AppSyncResolverEvent } from "aws-lambda/trigger/appsync-resolver";
import { AppSyncEventArguments, handleEvent } from "../../src/apiHandler";

class API extends AbstractApi {
  async callAPI(operation: OperationConfig, variables?: unknown): Promise<unknown> {
    return (
      await handleEvent({
        info: { fieldName: operation.name },
        arguments: variables,
      } as AppSyncResolverEvent<AppSyncEventArguments>)
    ).data;
  }

  async callYllapitoAPI(operation: OperationConfig, variables?: unknown): Promise<unknown> {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const payload = {
      info: { fieldName: operation.name },
      arguments: variables,
    } as AppSyncResolverEvent<AppSyncEventArguments>;
    return (await handleEvent(payload)).data;
  }
}

export const api = new API();
