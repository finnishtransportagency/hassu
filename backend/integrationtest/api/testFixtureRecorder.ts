import fs from "fs";
import { projektiDatabase } from "../../src/database/projektiDatabase";
import { DBProjekti } from "../../src/database/model";
import { DocumentClient } from "aws-sdk/lib/dynamodb/document_client";
import { localDocumentClient } from "../util/databaseUtil";

export enum FixtureName {
  NAHTAVILLAOLO = "NAHTAVILLAOLO",
}

export async function recordProjektiTestFixture(fixtureName: string | FixtureName, oid: string): Promise<void> {
  const dbProjekti = await projektiDatabase.loadProjektiByOid(oid);
  replaceFieldsByName(dbProjekti, "2020-01-01T00:00:00+03:00", "tuotu", "paivitetty");
  replaceFieldsByName(dbProjekti, "salt123", "salt");

  const oldValue = readRecord(fixtureName);
  const currentValue = JSON.stringify(dbProjekti, null, 2);
  // Prevent updating file timestamp so that running tests with "watch" don't get into infinite loop
  if (!oldValue || oldValue !== currentValue) {
    fs.writeFileSync(__dirname + "/records/" + fixtureName + ".json", currentValue);
  }
}

function readRecord(fixtureName: string | FixtureName) {
  const buffer = fs.readFileSync(__dirname + "/records/" + fixtureName + ".json");
  return buffer.toString("utf-8");
}

export async function useProjektiTestFixture(fixtureName: string | FixtureName): Promise<string> {
  const dbProjekti: Partial<DBProjekti> = JSON.parse(readRecord(fixtureName));
  // Write the copied projekti to archive table
  const putParams: DocumentClient.PutItemInput = {
    TableName: process.env.TABLE_PROJEKTI,
    Item: dbProjekti,
  };
  await localDocumentClient.put(putParams).promise();

  return dbProjekti.oid;
}

export function replaceFieldsByName(obj: unknown, value: unknown, ...fieldNames: string[]): void {
  Object.keys(obj).forEach(function (prop) {
    if (typeof obj[prop] == "object" && obj[prop] !== null) {
      replaceFieldsByName(obj[prop], value, ...fieldNames);
    } else if (fieldNames.indexOf(prop) >= 0) {
      obj[prop] = value;
    }
  });
}
