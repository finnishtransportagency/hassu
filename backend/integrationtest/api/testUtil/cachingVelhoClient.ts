import storage from "node-persist";
import dayjs from "dayjs";
import { velho, VelhoClient } from "../../../src/velho/velhoClient";
import * as sinon from "sinon";
import mocha from "mocha";
import { log } from "../../../src/logger";
import MockDate from "mockdate";

const CACHE_TTL_DAYS = 30;
const CACHE_REFRESH_DAYS = 29;

async function initCache() {
  return storage.init({
    ttl:  dayjs().add(CACHE_TTL_DAYS, "days").diff(dayjs()),
    parse: (str) => {
      // Korjataan stringiksi muutetut bufferit takaisin buffereiksi
      return JSON.parse(str, (k, v) => {
        if (v !== null && typeof v === "object" && "type" in v && v.type === "Buffer" && "data" in v && Array.isArray(v.data)) {
          return Buffer.from(v.data);
        }
        return v;
      });
    },
  });
}

const originalVelhoClient = new VelhoClient();

function cacheVelhoClientMethod<T>(velhoClientMethod: keyof typeof originalVelhoClient, originalFunc: (...argz: any[]) => Promise<T>) {
  let aStub: sinon.SinonStub;
  mocha.before(() => {
    aStub = sinon.stub(velho, velhoClientMethod);
  });
  mocha.beforeEach(async () => {
    if (!storage.removeExpiredItems) {
      doWithoutMockDate(async () => {
        await initCache();
      });
    }

    aStub.callsFake(
      async (...args: any[]): Promise<unknown> =>
        doWithoutMockDate(async () => {
          const cachekey = velhoClientMethod + JSON.stringify(args);
          const cachedResult = await storage.getDatum(cachekey);
          if (cachedResult && cachedResult.ttl) {
            const now = Date.now();
            const timeToTTL = (cachedResult.ttl || now) - now;
            const daysToExpiration = timeToTTL / 1000 / 3600 / 24;
            if (daysToExpiration < CACHE_REFRESH_DAYS) {
              try {
                const result = await originalFunc(...args);
                await storage.setItem(cachekey, result);
                return result;
              } catch (e) {
                // Ignore because there is a cached value anyway
                log.warn(e);
              }
            }
            // Käytetään cachetettua tulosta jos päivitettyä tulosta ei ole saatavilla
            return cachedResult.value;
          }
          const result = await originalFunc(...args);
          await storage.setItem(cachekey, result);
          return result;
        })
    );
  });
}

function doWithoutMockDate(func: () => void) {
  const savedDate = Date.now();
  MockDate.reset();
  try {
    return func();
  } finally {
    MockDate.set(savedDate);
  }
}

export function velhoCache(): void {
  cacheVelhoClientMethod("loadProjekti", (oid) => originalVelhoClient.loadProjekti(oid));
  cacheVelhoClientMethod("loadProjektiAineistot", (oid) => originalVelhoClient.loadProjektiAineistot(oid));
  cacheVelhoClientMethod("searchProjects", (term, requireExactMatch) => originalVelhoClient.searchProjects(term, requireExactMatch));
  cacheVelhoClientMethod("getAineisto", (oid) => originalVelhoClient.getAineisto(oid));
  cacheVelhoClientMethod("getLinkForDocument", (oid) => originalVelhoClient.getLinkForDocument(oid)); // Tuottaa expiroituvia linkkejä, jotka eivät oikeasti toimi. Jos haluat ladata aineistoja oikeasti, käytä getAineisto-metodia
}
