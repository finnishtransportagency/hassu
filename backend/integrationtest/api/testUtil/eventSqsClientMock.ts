import { SQSEvent, SQSRecord } from "aws-lambda/trigger/sqs";
import * as sinon from "sinon";
import { eventSqsClient } from "../../../src/scheduler/eventSqsClient";
import { handleEvent } from "../../../src/scheduler/sqsEventHandlerLambda";
import { Callback, Context } from "aws-lambda";
import mocha from "mocha";
import assert from "assert";
import { ScheduledEvent } from "../../../src/scheduler/scheduledEvent";
import { nyt, parseDate } from "../../../src/util/dateUtil";

export class EventSqsClientMock {
  fakeEventQueue: SQSEvent[] = [];
  sendScheduledEventStub!: sinon.SinonStub;

  constructor() {
    mocha.before(() => {
      this.sendScheduledEventStub = sinon.stub(eventSqsClient, "sendScheduledEvent");
    });
    mocha.beforeEach(() => {
      this.sendScheduledEventStub.callsFake(async (event) => {
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        this.fakeEventQueue.push({ Records: [{ body: JSON.stringify(event) } as SQSRecord] });
      });
    });
  }

  async processQueue(): Promise<void> {
    const deletedIndexes: number[] = [];
    await this.fakeEventQueue.reduce((promiseChain, event, index) => {
      return promiseChain.then(async () => {
        assert(event.Records && event.Records.length);
        const firstEvent = event.Records[event.Records.length - 1];
        const parsedFirstEvent: ScheduledEvent = JSON.parse(firstEvent.body);
        if (!parsedFirstEvent.date || (parsedFirstEvent.date && parseDate(parsedFirstEvent.date).isBefore(nyt()))) {
          await handleEvent(event, undefined as unknown as Context, undefined as unknown as Callback);
          deletedIndexes.push(index);
        }
      });
    }, Promise.resolve());
    this.fakeEventQueue = this.fakeEventQueue.filter((_event, index) => !deletedIndexes.includes(index));
  }
}
