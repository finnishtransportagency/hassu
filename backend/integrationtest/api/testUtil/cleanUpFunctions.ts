import * as API from "../../../../common/graphql/apiModel";
import { NahtavillaoloVaiheJulkaisu } from "../../../src/database/model";
import { cleanupAnyProjektiData } from "../testFixtureRecorder";

export function cleanupGeneratedIdAndTimestampFromFeedbacks(feedbacks?: API.Palaute[]): API.Palaute[] | undefined {
  return feedbacks
    ? feedbacks.map((palaute) => {
        cleanupAnyProjektiData(palaute);
        palaute.liite = palaute?.liite?.replace(palaute.id, "***unittest***");
        palaute.id = "***unittest***";
        palaute.vastaanotettu = "***unittest***";
        return palaute;
      })
    : undefined;
}

export function cleanupVuorovaikutusKierrosTimestamps(
  vuorovaikutusKierros: API.VuorovaikutusKierros | API.VuorovaikutusKierrosJulkaisu | API.VuorovaikutusKierrosJulkinen
): void {
  vuorovaikutusKierros.esittelyaineistot?.forEach((aineisto) => (aineisto.tuotu = "***unittest***"));
  vuorovaikutusKierros.suunnitelmaluonnokset?.forEach((aineisto) => (aineisto.tuotu = "***unittest***"));
  if (Object.keys(vuorovaikutusKierros).includes("__typename")) {
    (vuorovaikutusKierros as API.VuorovaikutusKierros).esittelyaineistot?.forEach(aineistoCleanupFunc);
    (vuorovaikutusKierros as API.VuorovaikutusKierros).suunnitelmaluonnokset?.forEach(aineistoCleanupFunc);
  }
}

function aineistoCleanupFunc(aineisto: API.Aineisto) {
  if (aineisto.tuotu) {
    aineisto.tuotu = "***unittest***";
  }
}

export function cleanupNahtavillaoloTimestamps(
  nahtavillaoloVaihe: API.NahtavillaoloVaiheJulkaisu | API.NahtavillaoloVaihe | NahtavillaoloVaiheJulkaisu | null | undefined
): API.NahtavillaoloVaiheJulkaisu | API.NahtavillaoloVaihe | NahtavillaoloVaiheJulkaisu | null | undefined {
  if (!nahtavillaoloVaihe) {
    return nahtavillaoloVaihe;
  }
  if (Object.keys(nahtavillaoloVaihe).includes("__typename")) {
    (nahtavillaoloVaihe as API.NahtavillaoloVaihe).aineistoNahtavilla?.forEach(aineistoCleanupFunc);
    (nahtavillaoloVaihe as API.NahtavillaoloVaihe).lisaAineisto?.forEach(aineistoCleanupFunc);
  }

  if ((nahtavillaoloVaihe as NahtavillaoloVaiheJulkaisu).hyvaksymisPaiva) {
    (nahtavillaoloVaihe as NahtavillaoloVaiheJulkaisu).hyvaksymisPaiva = "**unittest**";
  }

  if ((nahtavillaoloVaihe as NahtavillaoloVaiheJulkaisu).ilmoituksenVastaanottajat?.kunnat?.some((kunta) => kunta.messageId)) {
    (nahtavillaoloVaihe as NahtavillaoloVaiheJulkaisu).ilmoituksenVastaanottajat?.kunnat?.forEach((kunta) => {
      kunta.lahetetty = "***unittest***";
    });
  }

  if ((nahtavillaoloVaihe as NahtavillaoloVaiheJulkaisu).ilmoituksenVastaanottajat?.viranomaiset?.some((v) => v.messageId)) {
    (nahtavillaoloVaihe as NahtavillaoloVaiheJulkaisu).ilmoituksenVastaanottajat?.viranomaiset?.forEach((v) => {
      v.lahetetty = "***unittest***";
    });
  }

  const lisaAineistoParametrit = (nahtavillaoloVaihe as API.NahtavillaoloVaihe).lisaAineistoParametrit;
  if (lisaAineistoParametrit) {
    lisaAineistoParametrit.hash = "***unittest***";
    lisaAineistoParametrit.poistumisPaiva = "***unittest***";
    (nahtavillaoloVaihe as API.NahtavillaoloVaihe)["lisaAineistoParametrit"] = lisaAineistoParametrit;
  }
  return nahtavillaoloVaihe;
}

export function cleanupNahtavillaoloJulkaisuJulkinenTimestamps(
  nahtavillaoloVaihe: API.NahtavillaoloVaiheJulkaisuJulkinen | undefined | null
): API.NahtavillaoloVaiheJulkaisuJulkinen | undefined | null {
  if (nahtavillaoloVaihe) {
    nahtavillaoloVaihe.aineistoNahtavilla?.forEach((aineisto) => (aineisto.tuotu = "***unittest***"));
  }
  return nahtavillaoloVaihe;
}

export function cleanupHyvaksymisPaatosVaiheTimestamps(
  vaihe: API.HyvaksymisPaatosVaiheJulkaisu | API.HyvaksymisPaatosVaihe
): API.HyvaksymisPaatosVaiheJulkaisu | API.HyvaksymisPaatosVaihe {
  vaihe.aineistoNahtavilla?.forEach(aineistoCleanupFunc);
  vaihe.hyvaksymisPaatos?.forEach(aineistoCleanupFunc);

  if ((vaihe as API.HyvaksymisPaatosVaiheJulkaisu).ilmoituksenVastaanottajat?.kunnat?.some((kunta) => kunta.lahetetty)) {
    (vaihe as API.HyvaksymisPaatosVaiheJulkaisu).ilmoituksenVastaanottajat?.kunnat?.forEach((kunta) => {
      kunta.lahetetty = "***unittest***";
    });
  }

  if ((vaihe as API.HyvaksymisPaatosVaiheJulkaisu).ilmoituksenVastaanottajat?.viranomaiset?.some((v) => v.lahetetty)) {
    (vaihe as API.HyvaksymisPaatosVaiheJulkaisu).ilmoituksenVastaanottajat?.viranomaiset?.forEach((v) => {
      v.lahetetty = "***unittest***";
    });
  }

  return vaihe;
}

export function cleanupHyvaksymisPaatosVaiheJulkaisuJulkinenTimestamps(
  hyvaksymisPaatosVaihe: API.HyvaksymisPaatosVaiheJulkaisuJulkinen
): API.HyvaksymisPaatosVaiheJulkaisuJulkinen | undefined {
  if (!hyvaksymisPaatosVaihe) {
    return undefined;
  }
  hyvaksymisPaatosVaihe.aineistoNahtavilla?.forEach((aineisto) => (aineisto.tuotu = "***unittest***"));
  hyvaksymisPaatosVaihe.hyvaksymisPaatos?.forEach((aineisto) => (aineisto.tuotu = "***unittest***"));
  return hyvaksymisPaatosVaihe;
}

export function cleanupGeneratedIds<T extends Record<string, any>>(obj: T): Record<string, any> {
  return Object.keys(obj).reduce((cleanObj: Record<string, any>, key: string) => {
    const cleanedUpKey: string = key.replace(/[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}/g, "***unittest***");
    cleanObj[cleanedUpKey] = obj[key] as Record<string, any>;
    return cleanObj;
  }, {} as Record<string, any>);
}

export function cleanupNahtavillaUrlsInPDF(pdfText: string): string {
  return pdfText.replace(/http\S+nahtavillaolo/g, "***unittest***");
}
