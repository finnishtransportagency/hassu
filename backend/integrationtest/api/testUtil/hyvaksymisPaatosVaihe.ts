// @ts-nocheck

import { projektiDatabase } from "../../../src/database/projektiDatabase";
import { loadProjektiFromDatabase, loadProjektiJulkinenFromDatabase, testPublicAccessToProjekti } from "./tests";
import {
  HallintoOikeus,
  HyvaksymisPaatosVaiheTila,
  ProjektiKayttaja,
  Status,
  TilasiirtymaToiminto,
  TilasiirtymaTyyppi,
  VelhoAineisto,
  VelhoAineistoKategoria,
} from "../../../../common/graphql/apiModel";
import { UserFixture } from "../../../test/fixture/userFixture";
import { expect } from "chai";
import { api } from "../apiClient";
import { adaptAineistoToInput, expectToMatchSnapshot } from "./util";
import { apiTestFixture } from "../apiTestFixture";
import { cleanupHyvaksymisPaatosVaiheJulkaisuJulkinenTimestamps, cleanupHyvaksymisPaatosVaiheTimestamps } from "./cleanUpFunctions";
import dayjs from "dayjs";

export async function testHyvaksymisPaatosVaiheHyvaksymismenettelyssa(oid: string, userFixture: UserFixture): Promise<void> {
  const dbProjekti = await projektiDatabase.loadProjektiByOid(oid);
  const julkaisu = dbProjekti.nahtavillaoloVaiheJulkaisut[0];
  // Päättymispäivä alle vuosi menneisyyteen, jottei projekti mene epäaktiiviseksi
  julkaisu.kuulutusVaihePaattyyPaiva = dayjs().add(-2, "day").format();
  await projektiDatabase.updateNahtavillaoloVaiheJulkaisu(dbProjekti, julkaisu);

  userFixture.loginAsAdmin();
  await api.tallennaProjekti({
    oid,
    kasittelynTila: {
      hyvaksymispaatos: { asianumero: "asianro123", paatoksenPvm: "2022-06-09" },
    },
  });

  userFixture.loginAs(UserFixture.mattiMeikalainen);
  await loadProjektiFromDatabase(oid, Status.HYVAKSYMISMENETTELYSSA); // Verify status in yllapito

  // Verify status in public
  userFixture.logout();
  const publicProjekti = await loadProjektiJulkinenFromDatabase(oid, Status.HYVAKSYMISMENETTELYSSA);
  expect(publicProjekti.nahtavillaoloVaihe).not.to.be.undefined;
  expect(publicProjekti.nahtavillaoloVaihe.aineistoNahtavilla).to.be.undefined;
  userFixture.loginAs(UserFixture.mattiMeikalainen);
}

export async function testImportHyvaksymisPaatosAineistot(
  oid: string,
  velhoAineistoKategorias: VelhoAineistoKategoria[],
  projektiPaallikko: string
): Promise<void> {
  const lisaAineisto = velhoAineistoKategorias
    .reduce((documents, aineistoKategoria) => {
      aineistoKategoria.aineistot.forEach((aineisto) => documents.push(aineisto));
      return documents;
    }, [] as VelhoAineisto[])
    .sort((a, b) => a.oid.localeCompare(b.oid));

  await api.tallennaProjekti({
    oid,
    hyvaksymisPaatosVaihe: {
      hyvaksymisPaatos: adaptAineistoToInput([lisaAineisto[0]]),
      aineistoNahtavilla: adaptAineistoToInput(lisaAineisto.slice(2, 3)),

      ilmoituksenVastaanottajat: apiTestFixture.ilmoituksenVastaanottajat,
      kuulutusYhteystiedot: apiTestFixture.yhteystietoInputLista,
      kuulutusYhteysHenkilot: [projektiPaallikko],
      hallintoOikeus: HallintoOikeus.HAMEENLINNA,

      kuulutusPaiva: "2022-06-09",
      kuulutusVaihePaattyyPaiva: "2100-01-01",
    },
  });

  const projekti = await loadProjektiFromDatabase(oid, Status.HYVAKSYMISMENETTELYSSA);
  const hyvaksymisPaatosVaihe = projekti.hyvaksymisPaatosVaihe;
  const kasittelynTila = projekti.kasittelynTila;
  expectToMatchSnapshot("testImportHyvaksymisPaatosAineistot", {
    hyvaksymisPaatosVaihe,
    kasittelynTila,
  });
}

export async function testHyvaksymisPaatosVaiheApproval(
  oid: string,
  projektiPaallikko: ProjektiKayttaja,
  userFixture: UserFixture
): Promise<void> {
  userFixture.loginAsProjektiKayttaja(projektiPaallikko);
  await api.siirraTila({
    oid,
    tyyppi: TilasiirtymaTyyppi.HYVAKSYMISPAATOSVAIHE,
    toiminto: TilasiirtymaToiminto.LAHETA_HYVAKSYTTAVAKSI,
  });

  const projektiHyvaksyttavaksi = await loadProjektiFromDatabase(oid, Status.HYVAKSYMISMENETTELYSSA);
  expect(projektiHyvaksyttavaksi.hyvaksymisPaatosVaiheJulkaisut).to.have.length(1);
  expect(projektiHyvaksyttavaksi.hyvaksymisPaatosVaiheJulkaisut[0].tila).to.eq(HyvaksymisPaatosVaiheTila.ODOTTAA_HYVAKSYNTAA);

  await api.siirraTila({
    oid,
    tyyppi: TilasiirtymaTyyppi.HYVAKSYMISPAATOSVAIHE,
    toiminto: TilasiirtymaToiminto.HYVAKSY,
  });
  const projekti = await loadProjektiFromDatabase(oid, Status.HYVAKSYMISMENETTELYSSA);
  expectToMatchSnapshot("testHyvaksymisPaatosVaiheAfterApproval", {
    hyvaksymisPaatosVaihe: cleanupHyvaksymisPaatosVaiheTimestamps(projekti.hyvaksymisPaatosVaihe),
    hyvaksymisPaatosVaiheJulkaisut: projekti.hyvaksymisPaatosVaiheJulkaisut.map(cleanupHyvaksymisPaatosVaiheTimestamps),
  });

  // Move hyvaksymisPaatosVaiheJulkaisu into the past
  const dbProjekti = await projektiDatabase.loadProjektiByOid(oid);
  const julkaisu = dbProjekti.hyvaksymisPaatosVaiheJulkaisut[0];
  julkaisu.kuulutusVaihePaattyyPaiva = "2022-06-08";
  await projektiDatabase.updateHyvaksymisPaatosVaiheJulkaisu(dbProjekti, julkaisu);

  await testPublicAccessToProjekti(
    oid,
    Status.HYVAKSYTTY,
    userFixture,
    "HyvaksymisPaatosVaiheJulkinenAfterApproval",
    (projektiJulkinen) =>
      (projektiJulkinen.hyvaksymisPaatosVaihe = cleanupHyvaksymisPaatosVaiheJulkaisuJulkinenTimestamps(
        projektiJulkinen.hyvaksymisPaatosVaihe
      ))
  );
}
