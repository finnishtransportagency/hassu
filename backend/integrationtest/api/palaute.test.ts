/* tslint:disable:only-arrow-functions no-unused-expression */
import { describe, it } from "mocha";
import { FixtureName, useProjektiTestFixture } from "./testFixtureRecorder";
import { setupLocalDatabase } from "../util/databaseUtil";
import { deleteProjekti, tallennaLogo, verifyEmailsSent } from "./testUtil/tests";
import { UserFixture } from "../../test/fixture/userFixture";
import { userService } from "../../src/user";
import { api } from "./apiClient";
import { expectToMatchSnapshot, takeYllapitoS3Snapshot } from "./testUtil/util";
import { cleanupGeneratedIdAndTimestampFromFeedbacks } from "./testUtil/cleanUpFunctions";
import * as sinon from "sinon";
import { emailClient } from "../../src/email/email";

const oid = "1.2.246.578.5.1.2978288874.2711575506";

describe("Palaute", () => {
  let userFixture: UserFixture;
  let emailClientStub: sinon.SinonStub;

  before(async () => {
    userFixture = new UserFixture(userService);
    emailClientStub = sinon.stub(emailClient, "sendEmail");

    await setupLocalDatabase();
    try {
      await deleteProjekti(oid);
    } catch (_ignore) {
      // ignore
    }
    await useProjektiTestFixture(FixtureName.NAHTAVILLAOLO);
  });

  after(() => {
    userFixture.logout();
    sinon.restore();
  });

  it("should insert and manage feedback", async () => {
    userFixture.logout();
    const palauteId = await api.lisaaPalaute(oid, {
      etunimi: "Matti",
      sukunimi: "Meikalainen",
      puhelinnumero: "123456",
      sahkoposti: "test@vayla.fi",
      yhteydenottotapaPuhelin: true,
      yhteydenottotapaEmail: false,
      kysymysTaiPalaute: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
      liite: await tallennaLogo(),
    });

    userFixture.loginAs(UserFixture.mattiMeikalainen);
    const palautteet = await api.listaaPalautteet(oid);

    expectToMatchSnapshot("projekti palaute lisätty", cleanupGeneratedIdAndTimestampFromFeedbacks(palautteet));

    await api.otaPalauteKasittelyyn(oid, palauteId);

    const palautteetAfterFeedbackBeingHandled = await api.listaaPalautteet(oid);
    await expectToMatchSnapshot(
      "projekti palaute otettu käsittelyyn",
      cleanupGeneratedIdAndTimestampFromFeedbacks(palautteetAfterFeedbackBeingHandled)
    );
    verifyEmailsSent(emailClientStub);
    await takeYllapitoS3Snapshot(oid, "should insert and manage feedback", "palautteet");
  });
});
