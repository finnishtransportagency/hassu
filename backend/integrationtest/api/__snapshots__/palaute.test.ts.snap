// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palaute should insert and manage feedback 1`] = `
Object {
  "description": "projekti palaute lisätty",
  "mockedtime": "2023-01-01T00:00",
  "obj": Array [
    Object {
      "__typename": "Palaute",
      "etunimi": "",
      "id": "***unittest***",
      "kysymysTaiPalaute": "Nimetön palaute.",
      "liitteet": Array [],
      "oid": "1.2.246.578.5.1.2978288874.2711575506",
      "puhelinnumero": "",
      "sahkoposti": "",
      "sukunimi": "",
      "vastaanotettu": "***unittest***",
      "yhteydenottotapaEmail": true,
      "yhteydenottotapaPuhelin": true,
    },
    Object {
      "__typename": "Palaute",
      "etunimi": "Joku",
      "id": "***unittest***",
      "kysymysTaiPalaute": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
      "liitteet": Array [],
      "oid": "1.2.246.578.5.1.2978288874.2711575506",
      "puhelinnumero": "123456",
      "sahkoposti": "test@vayla.fi",
      "sukunimi": "Toinen",
      "vastaanotettu": "***unittest***",
      "yhteydenottotapaEmail": true,
      "yhteydenottotapaPuhelin": true,
    },
    Object {
      "__typename": "Palaute",
      "etunimi": "Matti",
      "id": "***unittest***",
      "kysymysTaiPalaute": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
      "liitteet": Array [
        Object {
          "__typename": "Liite",
          "liite": "/palautteet/***unittest***/logo.png",
          "skannausTulos": undefined,
        },
      ],
      "oid": "1.2.246.578.5.1.2978288874.2711575506",
      "puhelinnumero": "123456",
      "sahkoposti": "test@vayla.fi",
      "sukunimi": "Meikalainen",
      "vastaanotettu": "***unittest***",
      "yhteydenottotapaEmail": false,
      "yhteydenottotapaPuhelin": true,
    },
  ],
}
`;

exports[`Palaute should insert and manage feedback 2`] = `
Object {
  "description": "projekti palaute otettu käsittelyyn",
  "mockedtime": "2023-01-01T00:00",
  "obj": Array [
    Object {
      "__typename": "Palaute",
      "etunimi": "",
      "id": "***unittest***",
      "kysymysTaiPalaute": "Nimetön palaute.",
      "liitteet": Array [],
      "oid": "1.2.246.578.5.1.2978288874.2711575506",
      "puhelinnumero": "",
      "sahkoposti": "",
      "sukunimi": "",
      "vastaanotettu": "***unittest***",
      "yhteydenottotapaEmail": true,
      "yhteydenottotapaPuhelin": true,
    },
    Object {
      "__typename": "Palaute",
      "etunimi": "Joku",
      "id": "***unittest***",
      "kysymysTaiPalaute": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
      "liitteet": Array [],
      "oid": "1.2.246.578.5.1.2978288874.2711575506",
      "puhelinnumero": "123456",
      "sahkoposti": "test@vayla.fi",
      "sukunimi": "Toinen",
      "vastaanotettu": "***unittest***",
      "yhteydenottotapaEmail": true,
      "yhteydenottotapaPuhelin": true,
    },
    Object {
      "__typename": "Palaute",
      "etunimi": "Matti",
      "id": "***unittest***",
      "kysymysTaiPalaute": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
      "liitteet": Array [
        Object {
          "__typename": "Liite",
          "liite": "/palautteet/***unittest***/logo.png",
          "skannausTulos": undefined,
        },
      ],
      "oid": "1.2.246.578.5.1.2978288874.2711575506",
      "puhelinnumero": "123456",
      "sahkoposti": "test@vayla.fi",
      "sukunimi": "Meikalainen",
      "vastaanotettu": "***unittest***",
      "vastattu": true,
      "yhteydenottotapaEmail": false,
      "yhteydenottotapaPuhelin": true,
    },
  ],
}
`;

exports[`Palaute should insert and manage feedback 3`] = `
Array [
  Object {
    "subject": "Suunnitelmaan on tullut palautetta",
    "text": "Suunnitelmaan on tullut palautetta: https://localhost:3000/yllapito/projekti/1.2.246.578.5.1.2978288874.2711575506/suunnittelu",
    "to": "mikko.haapamaki@cgi.com",
  },
  Object {
    "subject": "Vahvistus palautteen jättämisestä Valtion liikenneväylien suunnittelu -järjestelmän kautta",
    "text": "Palaute vastaanotettu
01.01.2023 00:00
Nimi
Matti Meikalainen
Sähköposti
test@vayla.fi
Puhelinnumero
123456

Suunnitelman nimi
HASSU AUTOMAATTITESTIPROJEKTI1
Palaute
Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Sähköpostiviestin liitteet
logo.png
",
    "to": "test@vayla.fi",
  },
  Object {
    "subject": "Suunnitelmaan on tullut palautetta",
    "text": "Suunnitelmaan on tullut palautetta: https://localhost:3000/yllapito/projekti/1.2.246.578.5.1.2978288874.2711575506/suunnittelu",
    "to": "mikko.haapamaki@cgi.com",
  },
  Object {
    "subject": "Vahvistus palautteen jättämisestä Valtion liikenneväylien suunnittelu -järjestelmän kautta",
    "text": "Palaute vastaanotettu
01.01.2023 00:00
Nimi
Joku Toinen
Sähköposti
test@vayla.fi
Puhelinnumero
123456

Suunnitelman nimi
HASSU AUTOMAATTITESTIPROJEKTI1
Palaute
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.

",
    "to": "test@vayla.fi",
  },
  Object {
    "subject": "Suunnitelmaan on tullut palautetta",
    "text": "Suunnitelmaan on tullut palautetta: https://localhost:3000/yllapito/projekti/1.2.246.578.5.1.2978288874.2711575506/suunnittelu",
    "to": "mikko.haapamaki@cgi.com",
  },
]
`;

exports[`Palaute should insert and manage feedback 4`] = `
Object {
  "yllapito S3 files should insert and manage feedback": Object {
    "/***unittest***/logo.png": FileMetadata {
      "ContentDisposition": undefined,
      "ContentType": "image/png",
      "checksum": "ABC123",
    },
  },
}
`;
